# Instructions
This is an individual assignment. 

You are to use the attached diabetes dataset to train and evaluate any classifier for predicting the outcome (0 or 1). 

Before training, perform two transformations on the data:

    - Fill in the missing values. These will be zeros where a zero does not make sense. Research how to convert a zero into a NaN and then fill in the NaNs with the median value as in the lecture notes.
    
    - Standardize the data.

## Load the Dataset

In [17]:
import pandas as pd
diabetes_data = pd.read_csv('diabetes.csv')
diabetes_data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

## Describe the Dataset

In [18]:
diabetes_data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

## Inspect the Data

In [19]:
diabetes_data.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

## Replace Zeros with NaN, except for Outcome

In [20]:
import numpy as np
diabetes_data[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age']] = diabetes_data[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age']].replace(0, np.nan)
diabetes_data.head()



Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0          6.0    148.0           72.0           35.0      NaN  33.6   
1          1.0     85.0           66.0           29.0      NaN  26.6   
2          8.0    183.0           64.0            NaN      NaN  23.3   
3          1.0     89.0           66.0           23.0     94.0  28.1   
4          NaN    137.0           40.0           35.0    168.0  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

## Replace NaNs with median values

In [21]:
diabetes_data.fillna(diabetes_data.median(), inplace=True)
diabetes_data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0          6.0    148.0           72.0           35.0    125.0  33.6   
1          1.0     85.0           66.0           29.0    125.0  26.6   
2          8.0    183.0           64.0           29.0    125.0  23.3   
3          1.0     89.0           66.0           23.0     94.0  28.1   
4          4.0    137.0           40.0           35.0    168.0  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

## Separate features and target variable

In [22]:
from sklearn.preprocessing import StandardScaler
X = diabetes_data.iloc[:, :-1]
y = diabetes_data.iloc[:, -1]

# Standardize the features
scaler = StandardScaler()
X = scaler.fit_transform(X)

## Split data into training and testing sets

In [23]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Train a logistic regression classifier model

In [24]:
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression(random_state=42)
log_reg.fit(X_train, y_train)

LogisticRegression(random_state=42)

## Evaluate classifier on the testing set

In [25]:
from sklearn.metrics import accuracy_score
y_pred = log_reg.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.7597402597402597


## Check for Errors on Training & Testing Data

In [26]:
from sklearn.metrics import mean_squared_error
train_pred = log_reg.predict(X_train)
rmse = mean_squared_error(y_train, train_pred, squared=False)
print("Error on train data:", rmse)
test_pred = log_reg.predict(X_test)
rmse = mean_squared_error(y_test, test_pred, squared=False)
print("Error on test data:", rmse)

Error on train data: 0.47579865996117193
Error on test data: 0.4901629731627434


# Conclusion
My model has an accuracy score of approximately 76% which is not good but given the points I have laid out ahead, I guess it can be worked on and with. 

I think the reason for this poor performance comes from when I replaced zeros in the columns that had zeros with NaN and then replaced NaN with the median values. 

In my opinion, I think this messed up the realistic information we would have gotten from a non-tampered with data. 

Here, I took the risk of trading accuracy with uniformity of a dataset. 

When it came to checking for errors in the training and test data, both rmse values were high, again, pointing out the effect that came about when replacing zeros with the median values. 

Despite this, I think that this model can be improved with more research and more techniques applied.